### PR TITLE
Extend intent based navigation

### LINKED ISSUE #2184: Extend Intent based navigation
**Description**
- Add new structural way of calling intents.

- instead of `navigate('#?Intent=Sales-details?id=13&anotherId=20')` the new alternative is: 
 `navigateToIntent("Sales-details", {id:13, anotherId:20});`

-  new function definition:
`navigateToIntent(semanticSlug: string, params: object) {}`

- also add support for dynamic path segments.
   -  i.e.: `/projects/:project/settings` could be converted to `/projects/pr2/settings`

**Attachments**
[intent navigation](https://docs.luigi-project.io/docs/navigation-parameters-reference/?section=intentmapping)

### PR DESCRIPTION
**Pull request by ndricimrr**
**Description**
Introduces : `linkManager().navigateToIntent`
Offers an alternative way of navigating with intents. This involves specifying a semanticSlug and an object containing
 parameters. 

The new linkManager **`navigateToIntent`**  method internally generates a url of the form `#?intent=<semantic object>-<action>?<param_name>=<param_value>` through the given input arguments. This then follows a call to the original linkManager.navigate(...) function.
   * Consequentially the following calls shall have the exact same effect:
   * - `LuigiClient.linkManager().navigateToIntent('Sales-settings', {project: 'pr2', user: 'john'})`
   * - `LuigiClient.linkManager().navigate('/#?intent=Sales-settings?project=pr2&user=john')`
   
#### Changes proposed in this pull request:

- navigateToIntent() added
-  add support for dynamic path segments in intentMappings
i.e.: `/projects/:project/settings` is converted to `/projects/pr2/settings` if the {project:pr2} parameter is given through the intent.

**Related issue(s)**

Resolves #2184

### CODE REVIEW

File: docs/luigi-client-api.md [hunk 1]
```diff
@@ -946,60 +975,60 @@ Stores an item for a specific key.
...
--   `key` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** used to identify the value
+*   `key` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** used to identify the value
 
 ##### Examples
 
 ```javascript
 LuigiClient.storageManager().removeItem('keyExample').then((value) => console.log(value + ' just removed')
 ```
 
-Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)>** resolves an item just removed from storage. It will launch an error if storage is not supported or if you are using a Luigi reserved key.
```
Comment (alexandra-simeonova) [inline docs/luigi-client-api.md:998]: Looks like another change caused by the linting mentioned in #2212  🤔 But I don't think it causes any issues, right?

File: client/src/linkManager.js [hunk 2]
```diff
@@ -88,6 +88,37 @@ export class linkManager extends LuigiClientBase {
     helpers.sendPostMessageToLuigiCore(navigationOpenMsg);
   }
 
+  /**
+   * Offers an alternative way of navigating with intents. This involves specifying a semanticSlug and an object containing
+   * parameters.
+   * This method internally generates a url of the form `#?intent=<semantic object>-<action>?<param_name>=<param_value>` through the given
+   * input arguments. This then follows a call to the original linkManager.navigate(...) function.
+   * Consequentially the following calls shall have the exact same effect:
```
Comment (alexandra-simeonova) [inline client/src/linkManager.js:96]:
```suggestion
   * This method internally generates a URL of the form `#?intent=<semantic object>-<action>?<param_name>=<param_value>` through the given
   * input arguments. This then follows a call to the original `linkManager.navigate(...)` function.
   * Consequently, the following calls shall have the exact same effect:
```

File: client/src/linkManager.js [hunk 3]
```diff
@@ -88,6 +88,37 @@ export class linkManager extends LuigiClientBase {
...
+   * - linkManager().navigateToIntent('Sales-settings', {project: 'pr2', user: 'john'})
+   * - linkManager().navigate('/#?intent=Sales-settings?project=pr2&user=john')
+   * @param {string} semanticSlug concatenation of semantic object and action connected with a dash (-), i.e.: <semanticObject>-<action>
+   * @param {Object} params an object representing all the parameters passed, i.e.: {param1: '1', param2: 2, param3: 'value3'}
```
Comment (alexandra-simeonova) [inline client/src/linkManager.js:100]:
```suggestion
   * @param {string} semanticSlug concatenation of semantic object and action connected with a dash (-), i.e.: `<semanticObject>-<action>`
   * @param {Object} params an object representing all the parameters passed, i.e.: `{param1: '1', param2: 2, param3: 'value3'}`
```

File: core/src/utilities/helpers/routing-helpers.js [hunk 4]
```diff
@@ -416,6 +418,39 @@ class RoutingHelpersClass {
...
+  /**
+   * This function takes a path which contains dynamic parameters and a list parameters and replaces the dynamic parameters
+   * with the given parameters if any. The input path remains unchanged if the parameters list
+   * does not contain the respective dynamic parameter name.
+   * e.g.:
+   * Assume either of these two calls are made:
+   * 1. linkManager().navigateToIntent('Sales-settings', {project: 'pr2', user: 'john'})
+   * 2. linkManager().navigate('/#?intent=Sales-settings?project=pr2&user=john')
+   * For both 1. and 2. calls the following dynamic input path: `/projects/:project/details/:user`
+   * is resolved through this method to `/projects/pr2/details/john`
```
Comment (alexandra-simeonova) [inline core/src/utilities/helpers/routing-helpers.js:431]:
```suggestion
   * Assume either of these two calls are made:
   * 1. `linkManager().navigateToIntent('Sales-settings', {project: 'pr2', user: 'john'})`
   * 2. `linkManager().navigate('/#?intent=Sales-settings?project=pr2&user=john')`
   * For both 1. and 2., the following dynamic input path: `/projects/:project/details/:user`
   * is resolved through this method to `/projects/pr2/details/john`
```

File: core/src/utilities/helpers/routing-helpers.js [hunk 5]
```diff
@@ -416,6 +418,39 @@ class RoutingHelpersClass {
...
+   *
+   * @param {string} path the path containing potential dynamic parameter
+   * @param {Object} parameters a list of objects consisting in passed parameters
```
Comment (alexandra-simeonova) [inline core/src/utilities/helpers/routing-helpers.js:434]:
```suggestion
   * @param {string} path the path containing the potential dynamic parameter
   * @param {Object} parameters a list of objects consisting of passed parameters
```

File: client/src/linkManager.js [hunk 6]
```diff
@@ -88,6 +88,37 @@ export class linkManager extends LuigiClientBase {
...
+   * - linkManager().navigateToIntent('Sales-settings', {project: 'pr2', user: 'john'})
+   * - linkManager().navigate('/#?intent=Sales-settings?project=pr2&user=john')
+   * @param {string} semanticSlug concatenation of semantic object and action connected with a dash (-), i.e.: `<semanticObject>-<action>`
+   * @param {Object} params an object representing all the parameters passed, i.e.: `{param1: '1', param2: 2, param3: 'value3'}`
+   * @example
+   * LuigiClient.linkManager().navigateToIntent('Sales-settings', {project: 'pr2', user: 'john'})
+   */
+  navigateToIntent(semanticSlug, params) {
+    let newPath = '#?intent=';
+    newPath += semanticSlug;
```
Comment (stanleychh) [inline client/src/linkManager.js:106]: Do we have a potential XSS attack possibility here?

File: docs/luigi-client-api.md (hunk 1 above)
Comment (ndricimrr) [reply docs/luigi-client-api.md:998]: Okay, I hadn't seen this one change, but it does seem to actually break the links. Re-opening the issue #2212

**Review by alexandra-simeonova — APPROVED** [review]

File: core/test/services/routing.spec.js [hunk 7]
```diff
@@ -205,22 +189,69 @@ describe('Routing', function () {
...
+    });
+
+    it('input path not changed when there are no parameters defined', () => {
+      const path = '/projects/:component/details/:componentId/view/:componentCount/show';
+      const parameters = undefined;
+      const actual = RoutingHelpers.resolveDynamicIntentPath(path, parameters);
+      const expected = '/projects/:component/details/:componentId/view/:componentCount/show';
+      assert.equal(actual, expected);
+    });
+
```
[the remark this answers is not in the document]
Comment (ndricimrr) [reply core/test/services/routing.spec.js:237]: Added the test

File: docs/luigi-client-api.md (hunk 1 above)
Comment (ndricimrr) [reply docs/luigi-client-api.md]: Resolved it by manually updating docu.

File: client/src/linkManager.js (hunk 6 above)
Comment (ndricimrr) [reply client/src/linkManager.js:106]: This is currently not possible to happen I think. I think this was already prevented when we added intent-based navigation the first time around. 
[In this line](https://github.com/SAP/luigi/blob/master/core/src/utilities/helpers/routing-helpers.js#L346) we check for the semantic-object with an alphanumeric regexp. 
While the parameters we also sanitize them when we build the node. Like [here](https://github.com/SAP/luigi/blob/master/core/src/utilities/helpers/routing-helpers.js#L81) and [here](https://github.com/SAP/luigi/blob/master/core/src/navigation/services/navigation.js#L165) . Basically just tried it out and they get sanitized properly.

**Review by stanleychh — APPROVED** [review]
Looks good 👍  Thanks.

**Pull request merged**